common/src/main/scala/explore/AppMain.scala

```diff
@@ -65,7 +69,21 @@ trait AppMain extends IOApp {
         }
       }
 
+    def refreshToken(expiration:       Instant, v: View[RootModel]): IO[UserVault] =
+      IO(Instant.now).flatMap { n =>
+        val sleepTime = SSOClient.refreshTimoutDelta.max(
+          (n.until(expiration, ChronoUnit.SECONDS).seconds - SSOClient.refreshTimoutDelta)
+        )
+        IO.timer.sleep(sleepTime)
+      } *> SSOClient
+        .whoami[IO](IO.fromFuture)
+        .flatTap((u: UserVault) => v.zoom(RootModel.vault).set(u))
+
+    def repeatTokenRefresh(expiration: Instant, v: View[RootModel]): IO[Unit]      =
+      refreshToken(expiration, v).flatMap(u => repeatTokenRefresh(u.expiration, v))
```

Comment:
This is a very small thing: `refreshToken` and `repeatTokenRefresh` don't need to access all of `RootModel`, they could just receive a `View[UserVault]` and have the `zoom` in the caller.

Reply:
Good call, You're right we may want to focus on the relevant field.